torch/csrc/jit/codegen/cuda/scheduler.cpp

```diff
@@ -670,6 +885,461 @@ void scheduleReduction(
+      [](TensorView* left, TensorView* right) {
+        return left->name() > right->name();
+      });
+  return duplicate_tv;
+}
+
+//! Find all TensorViews that require inline ComputeAt
+//! to avoid non-static allocation error
+std::vector<TensorView*> findTensorViewsToComputeAtInline(
+    Fusion* fusion,
+    const std::vector<TensorView*>& other_tv) {
+  std::vector<TensorView*> computeAt_inline_tv;
+  for (auto tv : other_tv) {
+    if (!fusion->hasInput(tv) && !fusion->hasOutput(tv)) {
+      auto domain = tv->domain()->domain();
+      bool constant_allocation = true;
+      for (size_t axis = tv->getThisComputeAtAxis(); axis < domain.size();
+           ++axis) {
+        if (!domain[axis]->isBroadcast() && !domain[axis]->isReduction()) {
+          constant_allocation &= domain[axis]->isConstScalar();
+        }
+      }
+      if (!constant_allocation && tv->getMemoryType() == MemoryType::Local) {
+        computeAt_inline_tv.push_back(tv);
+      }
+    }
+  }
+  return computeAt_inline_tv;
+}
+
+//! Place all cache TensorViews in Shared Memory
+//! All point-wise TensorViews inherit shared memory from their parents
+void setupSharedMemory(
+    Fusion* fusion,
+    const std::vector<TensorView*>& cache_tv) {
+  std::vector<TensorView*> stack(cache_tv.begin(), cache_tv.end());
+  while (!stack.empty()) {
+    auto tensor = stack.back();
+    stack.pop_back();
+    if (!fusion->hasOutput(tensor) && !fusion->hasInput(tensor)) {
+      tensor->setMemoryType(MemoryType::Shared);
+      for (auto expr : fusion->unordered_uses(tensor)) {
+        if (isPointwiseOp(expr)) {
+          auto output = expr->output(0)->as<TensorView>();
+          stack.push_back(output);
+        }
+      }
+    }
+  }
+}
+
+void organizeAxes(
+    const std::vector<TensorView*>& reduction_tv,
+    const std::vector<TensorView*>& all_tv) {
+  // Determine merged reduction axis position
+  auto findMergedReductionAxis = [](TensorView* reduction_tv) {
+    int merged_reduction_axis = -1;
+    auto domain = reduction_tv->domain()->domain();
+    for (size_t axis = 0; axis < domain.size(); ++axis) {
+      if (domain[axis]->isReduction()) {
+        TORCH_INTERNAL_ASSERT(merged_reduction_axis == -1);
+        merged_reduction_axis = axis;
+      }
+    }
+    return merged_reduction_axis;
+  };
+
+  auto first_reduction_tv = reduction_tv.front();
+  auto root_domain = first_reduction_tv->getRootDomain();
+  int merged_reduction_axis = -1;
+
+  // Find reduction axes positions
+  std::vector<int> reduction_axes;
+  for (size_t axis = 0; axis < root_domain.size(); ++axis) {
+    if (root_domain[axis]->isReduction()) {
+      reduction_axes.push_back(axis);
+    }
+  }
+
+  // Coalese reduction axes together
+  for (auto tv : all_tv) {
+    const int kOuterAxis = reduction_axes.front();
+    for (int idx = 0; idx < reduction_axes.size() - 1; ++idx) {
+      int inner_axis = reduction_axes[idx + 1] - idx;
+      tv->merge(kOuterAxis, inner_axis);
+    }
+  }
```

Comment:
Are you content with the reduction axis moving to the left when getting merged, or should we support a case where we actually move it to the right? Don't think it would be too hard to modify merge so it moves to the right if you think that would be better.
Just wondering about the case where we have:
[I, R, I, R]
We would flag the reduction axis as being inner most, but we would end up with an iter axis to the right of the reduction. Would the scheduling logic get confused about this?

Reply:
Scheduling configuration is determined by the root domain. There is also a check to move the merged reduction axis to the inner-most position later in the function.